firestore-translate-text/__tests__/functions.test.ts

```diff
@@ -0,0 +1,384 @@
+        { changed: 123 },
+        "translations/id1"
+      );
+      documentChange = functionsTest.makeChange(beforeSnapshot, afterSnapshot);
+
+      const callResult = await wrappedFunction(documentChange);
+
+      expect(callResult).toBeUndefined();
+
+      expect(mockConsoleLog).toHaveBeenCalledWith(
+        "Document was created without an input string, no processing is required"
+      );
+
+      expect(mockTranslateClassMethod).not.toHaveBeenCalled();
+      expect(mockFirestoreUpdate).not.toHaveBeenCalled();
+    });
+
+    test("function exits early if input & output fields are the same", async () => {
+      jest.resetModules();
+
+      restoreEnv();
+      restoreEnv = mockedEnv({
+        ...defaultEnvironment,
+        INPUT_FIELD_NAME: "input",
+        OUTPUT_FIELD_NAME: "input",
+      });
+
+      wrappedFunction = functionsTest.wrap(
+        require("../functions/src").fstranslate
+      );
+
+      const callResult = await wrappedFunction(documentChange);
+
+      expect(callResult).toBeUndefined();
+      expect(mockConsoleError).toHaveBeenCalledWith(
+        "The `Input` and `Output` field names must be different for this extension to function correctly"
+      );
+    });
+
+    test("function exits early if input field is a translation output path", async () => {
+      jest.resetModules();
+
+      restoreEnv();
+      restoreEnv = mockedEnv({
+        ...defaultEnvironment,
+        INPUT_FIELD_NAME: "output.en",
+        OUTPUT_FIELD_NAME: "output",
+      });
+
+      wrappedFunction = functionsTest.wrap(
+        require("../functions/src").fstranslate
+      );
+
+      const callResult = await wrappedFunction(documentChange);
+
+      expect(callResult).toBeUndefined();
+      expect(mockConsoleError).toHaveBeenCalledWith(
+        "The `Input` field name must not be the same as an `Output` path for this extension to function correctly"
+      );
+    });
+
+    test("function updates translation document with translations", async () => {
+      mockFirestoreUpdate.mockClear();
+      mockConsoleLog.mockClear();
+
+      await wrappedFunction(documentChange);
+
+      // confirm Google Translate API was called
+      expect(mockTranslateClassMethod).toHaveBeenCalledWith("hello", "en");
+      expect(mockTranslateClassMethod).toHaveBeenCalledWith("hello", "es");
+      expect(mockTranslateClassMethod).toHaveBeenCalledWith("hello", "fr");
+      expect(mockTranslateClassMethod).toHaveBeenCalledWith("hello", "de");
+
+      // confirm document update was called
+      expect(mockFirestoreUpdate).toHaveBeenCalledWith(
+        defaultEnvironment.OUTPUT_FIELD_NAME,
+        testTranslations
+      );
+
+      // confirm logs were printed
+      Object.keys(testTranslations).forEach((language) => {
+        // logs.translateInputString
+        expect(mockConsoleLog).toHaveBeenCalledWith(
+          `Translating string: 'hello' into language(s): '${language}'`
```

Comment:
I really like that you're verifying the the logs were printed b/c we and our users will rely on this to diagnose issues. I'm concerned about the fragility of this though if we ever want to redo our messages.

**Random thought:** Perhaps refactor the logging to include a code in the front and pass in a structured object separately as part of `console.log`? The test assertions could just check the presence of the code and inspect structured data which will be more typesafe. I'm open to other ideas as well.